src/cvrp/ecvrp.py

```diff
@@ -80,58 +90,415 @@ def get_roads(self) -> tuple[tuple[int, ...], ...]:
+            for index_point, point in enumerate(road):
+                if (index_point == 0) and (index_road == 0):
+                    fusion_solution.append(point)
+                elif index_point != 0:
+                    # the goal is to not have 0 in duplicate
+                    # which is all the time in index 0 for each road
+                    fusion_solution.append(point)
+        return fusion_solution
+
+    def __pull_off_chargers(self, road: list[int]) -> list[int]:
+        """
+        Remove all the chargers from a solution.
+
+        This function is helping the mutation process by removing all the chargers
+        of a road. It will retrun the road without it.
+        """
+        for point in road:
+            if self.__instance.is_charger(point):
+                road.remove(point)
+        return road
+
+    def __reversed_content(self, road: list[int], index1: int, index2: int) -> list[int]:
+        """Reverse the content of a road between the point 1 and 2 (included)."""
+        size = len(road)
+
+        if index1 < index2:
+            index_min = index1
+            index_max = index2
+        else:
+            index_min = index2
+            index_max = index1
+
+        begining = road[0:index_min]
+        middle = road[index_min:(index_max+1)]
+        middle.reverse()
+        ending = road[(index_max+1):size]
+
+        reversed_road = begining + middle + ending
+
+        return reversed_road
+
+    def __two_opt(self, road: list[int]) -> list[int]:
+        """
+        2-opt variance algorithm.
+
+        Algorithm which find the best mutation for this road by calculating
+        the distance for each mutation of road. The shortest distance will
+        determine the best muatation road.
+        """
+        best_distance = self._compute_road_fitness(road)
+        best_road = road
+
+        depot = self.__instance.get_depot()
+
+        road.remove(depot)  # remove the first depot
+        road.remove(depot)  # remove the last depot
+
+        for index_point1 in range(len(road)):
+            for index_point2 in range(len(road)):
+                if index_point1 != index_point2:
+                    tmp_road = self.__reversed_content(road, index_point1, index_point2)
+                    tmp_road.insert(0, depot)  # add the depot at the begining
+                    tmp_road.append(depot)  # add the depot at the end
+                    distance = self._compute_road_fitness(tmp_road)
+                    if distance < best_distance:
+                        best_distance = distance
+                        best_road = tmp_road
+
+        return best_road
+
+    def delete_empty_road(self, solution: list[list[int]]) -> list[list[int]]:
+        """Delete empty road which are represented by [0, 0] if depot is 0."""
+        depot = self.__instance.get_depot()
+        for road in solution:
+            if road == [depot, depot]:
+                solution.remove(road)
```

Comment:
The `remove` method iterate over the list to find the relevant entry and then remove it. Using `pop` would be more efficient here. (Both are O(n) but pop has a lower factor) You might want to do a little of testing before applying this change.

Reply:
But I can't, pop works with indexes, and we don't know where is the empty road, it can be everywhere.
Am I making a mistake somewhere?

Reply:
If you use enumerate you can access both the index and the road.

Reply:
It's true

Reply:
done